### PR TITLE
include cost of transaction's requested loaded accounts size in cost model

### PR DESCRIPTION
**Pull request by taozhu-chicago**
#### Problem

There is cost associated with amount memory space transactions request to load accounts into. #30659 formulated cost into CUs for base fee calculation. Can use same formula in cost model, also behind `include_loaded_accounts_data_size_in_fee_calculation` feature gate

#### Summary of Changes
- Add `loaded_accounts_data_size_cost` to cost_model. When feature gate `include_loaded_accounts_data_size_in_fee_calculation` is activated, additional CUs for loaded_accounts_data_size will be contributed to overall transaction cost;

Note: Behind feature Gate Issue #30657

### CODE REVIEW

File: runtime/src/cost_model.rs [hunk 1]
```diff
@@ -163,16 +169,16 @@ impl CostModel {
         }
 
         // calculate bpf cost based on compute budget instructions
-        let mut budget = ComputeBudget::default();
+        let mut compute_budget = ComputeBudget::default();
 
-        // Starting from v1.15, cost model uses compute_budget.set_compute_unit_limit to
+        // Starting from v1.15, cost model uses compute_compute_budget.set_compute_unit_limit to
```
Comment (alessandrod) [inline runtime/src/cost_model.rs:174]: nit: compute_compute_budget

File: runtime/src/cost_model.rs [hunk 2]
```diff
@@ -163,16 +169,16 @@ impl CostModel {
...
         // measure bpf_costs (code below), vs earlier versions that use estimated
         // bpf instruction costs. The calculated transaction costs are used by leaders
         // during block packing, different costs for same transaction due to different versions
-        // will not impact consensus. So for v1.15+, should call compute budget with
+        // will not impact consensus. So for v1.15+, should call compute compute_budget with
```
Comment (alessandrod) [inline runtime/src/cost_model.rs:178]: here too, find & replace :P

**Review by alessandrod — COMMENTED** [review]
lgtm! see comment nits

File: runtime/src/cost_model.rs (hunk 1 above)
Comment (taozhu-chicago) [reply runtime/src/cost_model.rs:174]: Ooops, search/replace beyond scope :) Thanks for pointing out, fixing it

File: runtime/src/cost_model.rs [hunk 3]
```diff
@@ -6,15 +6,16 @@
 //!
 
 use {
-    crate::block_cost_limits::*,
+    crate::{bank::Bank, block_cost_limits::*},
```
Comment (apfitzge) [inline runtime/src/cost_model.rs:9]: I'm hesistant on introducing this, it's creating a cyclic dependence between our modules.

```mermaid
flowchart TD

bank;
cost_tracker;
cost_model;

bank --> cost_tracker;
cost_tracker --> cost_model;
cost_model --> bank;
```

There's not a technical reason we can't do this, but imo it points to `Bank::calculate_loaded_accounts_data_size_cost` maybe being in the "wrong" spot.

Would that function make more sense to be on `ComputeBudget` or `CostModel`?

Comment (taozhu-chicago) [reply runtime/src/cost_model.rs:9]: Good call! Was being lazy to simply want to reuse the function already in `bank`. Should move it to `CostModel`. 

Also this points out an unnecessary dependency between CostTrack and CostModel - CostTracker doesn't need to know CostModel, it only needs TransactionCost. Will clean it up in a separate PR.

File: runtime/src/cost_model.rs [hunk 4]
```diff
@@ -104,9 +122,16 @@ impl CostModel {
 
         // if tx contained user-space instructions and a more accurate estimate available correct it
         if bpf_costs > 0 && result.is_ok() {
-            bpf_costs = budget.compute_unit_limit
+            bpf_costs = compute_budget.compute_unit_limit
         }
 
+        tx_cost.loaded_accounts_data_size_cost =
```
[the remark this answers is not in the document]
Comment (taozhu-chicago) [reply runtime/src/cost_model.rs]: yea, merged #31976, adjusted logic to not count loaded_account_size_cost if tx isn't going to reach runtime

File: runtime/src/cost_model.rs [hunk 5]
```diff
@@ -110,6 +129,13 @@ impl CostModel {
                 if bpf_costs > 0 {
                     bpf_costs = budget.compute_unit_limit
                 }
+                loaded_accounts_data_size_cost = if feature_set
+                    .is_active(&include_loaded_accounts_data_size_in_fee_calculation::id())
+                {
+                    Self::calculate_loaded_accounts_data_size_cost(&compute_budget)
+                } else {
```
Comment (apfitzge) [inline runtime/src/cost_model.rs:136]: nit: the else isn't needed here since it's initialized to 0u64

File: runtime/src/cost_model.rs [hunk 6]
```diff
@@ -507,10 +533,124 @@ mod tests {
         let expected_execution_cost = BUILT_IN_INSTRUCTION_COSTS
             .get(&system_program::id())
             .unwrap();
+        // feature `include_loaded_accounts_data_size_in_fee_calculation` enabled, using
+        // default loaded_accounts_data_size_limit
+        let expected_loaded_accounts_data_size_cost = (64 * 1024 * 1024) / (32 * 1024) * 8;
```
Comment (apfitzge) [inline runtime/src/cost_model.rs:538]: seems like this can/should be calculated using some constants?

File: runtime/src/cost_model.rs [hunk 7]
```diff
@@ -634,14 +634,8 @@ mod tests {
...
         let expected_account_cost = WRITE_LOCK_UNITS * 2;
-        let expected_execution_cost = BUILT_IN_INSTRUCTION_COSTS
-            .get(&system_program::id())
-            .unwrap()
-            + BUILT_IN_INSTRUCTION_COSTS
-                .get(&compute_budget::id())
-                .unwrap();
-        // feature `include_loaded_accounts_data_size_in_fee_calculation` is not enabled, accounts data
-        // size limit is set.
+        // with features all diabled, builtins and loaded account size don't cost CU
```
Comment (apfitzge) [inline runtime/src/cost_model.rs:642]: typo:
```suggestion
        // with features all disabled, builtins and loaded account size don't cost CU
```

**Comment by apfitzge** [comment]
PR description be updated. This is related to #30657, but with the "fixes" it will close that issue, which I don't think we want until the feature is actually activated everywhere

**Comment by taozhu-chicago** [comment]
> PR description be updated. This is related to #30657, but with the "fixes" it will close that issue, which I don't think we want until the feature is actually activated everywhere

Yea, that can be confusing. Changed to "Behind Feature gate Issue...". Also commented on the feature issue to include cost model part of impact when activating gate

**Review by apfitzge — APPROVED** [review]
lgtm

**Comment by yihau** [comment]
Hi there. it seems that this one make our system performance test become so bad.

```
mean_tps: 5435
max_tps: 8062
mean_confirmation_ms: 694
max_confirmation_ms: 1421
99th_percentile_confirmation_ms: 1065
max_tower_distance: 31
last_tower_distance: 31
slots_per_second: 2.823
```

normally, it should be
```
mean_tps: 25155
max_tps: 97440
mean_confirmation_ms: 827
max_confirmation_ms: 4462
99th_percentile_confirmation_ms: 3105
max_tower_distance: 46
last_tower_distance: 31
slots_per_second: 2.636
```

**Comment by taozhu-chicago** [comment]
In the test where all features are enabled, it is expected that transactions to use instruction `compute_budget::set_loaded_account_data_size_limit` to set bytes it needs to load all accounts. Otherwise MAX=64M will be used, which will use up a lot of block limit with this PR.

**Comment by apfitzge** [comment]
> In the test where all features are enabled, it is expected that transactions to use instruction compute_budget::set_loaded_account_data_size_limit to set bytes it needs to load all accounts. Otherwise MAX=64M will be used, which will use up a lot of block limit with this PR.

So our choice is to either disable the feature for bench-tps, or to set the data size reasonably on the txs.
I'd think that the second option is better.

**Comment by taozhu-chicago** [comment]
It also leads to question "how many bytes should I request for my transactions"? Need to add support to simulator or RPC for end users to query accounts size. Bump it up from my todo list.

**Pull request merged**